web-vr/hello-world/demo/demo-vr.js

```diff
@@ -0,0 +1,203 @@
+
+'use strict';
+
+/* eslint-env es6 */
+
+class DemoVR extends Demo {
+  constructor () {
+    super();
+
+    this._disabled = false;
+    if (typeof VRFrameData === 'undefined') {
+      this._disabled = true;
+      this._showError();
+      return;
+    }
+
+    this._button = undefined;
+    this._vr = {
+      display: null,
+      frameData: new VRFrameData()
+    };
+
+    this._addVREventListeners();
+    this._getDisplays();
+  }
+
+  _addVREventListeners () {
+    window.addEventListener('vrdisplayactivate', evt => {
+      evt.preventDefault();
+      this._activateVR();
+    });
+
+    window.addEventListener('vrdisplaydeactivate', evt => {
+      evt.preventDefault();
+      this._deactivateVR();
+    });
+  }
+
+  _getDisplays () {
+    return navigator.getVRDisplays().then(displays => {
+      if (displays.length === 0) {
+        return;
+      }
+
+      this._vr.display = displays[0];
+      this._vr.depthNear = DemoVR.CAMERA_SETTINGS.near;
+      this._vr.depthFar = DemoVR.CAMERA_SETTINGS.far;
+
+      if (!this._vr.display.capabilities.canPresent) {
+        return this._showNoPresent();
+      }
+
+      this._createPresentationButton();
+    });
+  }
+
+  _showNoPresent () {
+    console.warn(`Unable to present with this device ${this._vr.display}`);
+  }
+
+  _showError () {
+    console.error('WebVR not supported');
+  }
+
+  _createPresentationButton () {
+    this._button = document.createElement('button');
+    this._button.classList.add('vr-toggle');
+    this._button.addEventListener('click', _ => {
+      this._toggleVR();
+    });
+    document.body.appendChild(this._button);
+
+    this._updateButtonLabel();
+  }
+
+  _updateButtonLabel () {
+    this._button.textContent = this._vr.display.isPresenting ?
+        'Disable VR' :
+        'Enable VR';
+  }
+
+  _deactivateVR () {
+    if (!this._vr.display) {
+      return;
+    }
+
+    this._vr.display.exitPresent();
+    this._updateButtonLabel();
+    return Promise.resolve();
+  }
+
+  _activateVR () {
+    if (!this._vr.display) {
+      return;
+    }
+
+    this._vr.display.requestPresent([{
+      source: this._renderer.domElement
+    }])
+    .then(_ => {
+      this._updateButtonLabel();
+    })
+    .catch(e => {
+      console.error(`Unable to init VR: ${e}`);
+    });
+  }
+
+  _toggleVR () {
+    if (this._vr.display.isPresenting) {
+      return this._deactivateVR();
+    }
+
+    return this._activateVR();
+  }
+
+  _render () {
+    if (this._disabled || !(this._vr.display && this._vr.display.isPresenting)) {
+      // Ensure that we switch everything back to auto for non-VR mode.
+      this._onResize();
+      this._renderer.autoClear = true;
+      this._scene.matrixAutoUpdate = true;
+
+      return super._render();
+    }
+
+    const EYE_WIDTH = this._width * 0.5;
+    const EYE_HEIGHT = this._height;
+
+    // Get all the latest data from the VR headset and dump it into frameData.
+    this._vr.display.getFrameData(this._vr.frameData);
```

Comment:
I don't see a submitFrame call... isn't that needed to make the matrices update? I believe that this is somewhere that the polyfill and extension differ from the spec, so check in a real device.